libc/stdio_scanf.c

```diff
@@ -0,0 +1,3701 @@
+
+	TEST_ASSERT_EQUAL_DOUBLE(expinf, inf);
+	TEST_ASSERT_EQUAL_DOUBLE(HUGE_VAL * -1, neginf);
+	TEST_ASSERT_EQUAL_DOUBLE(NAN, this_nan);
+
+	rewind(rd);
+
+	TEST_ASSERT_EQUAL_INT(3, fscanf(rd, ext, &inf, &neginf, &this_nan));
+
+	TEST_ASSERT_EQUAL_DOUBLE(expinf, inf);
+	TEST_ASSERT_EQUAL_DOUBLE(HUGE_VAL * -1, neginf);
+	TEST_ASSERT_EQUAL_DOUBLE(NAN, this_nan);
+
+	TEST_ASSERT_EQUAL_INT(
+		3, vsscanf_wrapper(tested_str, ext, &inf, &neginf, &this_nan));
+
+	TEST_ASSERT_EQUAL_DOUBLE(expinf, inf);
+	TEST_ASSERT_EQUAL_DOUBLE(HUGE_VAL * -1, neginf);
+	TEST_ASSERT_EQUAL_DOUBLE(NAN, this_nan);
+
+	TEST_ASSERT_EQUAL_INT(3, sscanf(tested_str, ext, &inf, &neginf, &this_nan));
+
+	TEST_ASSERT_EQUAL_DOUBLE(expinf, inf);
+	TEST_ASSERT_EQUAL_DOUBLE(HUGE_VAL * -1, neginf);
+	TEST_ASSERT_EQUAL_DOUBLE(NAN, this_nan);
+}
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+TEST_SETUP(stdio_scanf_str_chr_ptr) { }
+
+TEST_TEAR_DOWN(stdio_scanf_str_chr_ptr) { unlink(std_path); }
+
+TEST(stdio_scanf_str_chr_ptr, c)
+{
+
+	char tested_str[5], tested_str2[5];
+	char chr;
+	char T = 'T';
+	char tstr[] = "Test";
+
+	sprintf(tested_str, "%c", T);
+	create_file(std_path, tested_str);
+	rd = fopen(std_path, "r");
+
+	TEST_ASSERT_EQUAL_INT(1, vfscand_wrapper(rd, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(T, chr);
+
+	rewind(rd);
+
+	TEST_ASSERT_EQUAL_INT(1, fscanf(rd, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(T, chr);
+
+	TEST_ASSERT_EQUAL_INT(1, vsscanf_wrapper(tested_str, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(T, chr);
+
+	TEST_ASSERT_EQUAL_INT(1, sscanf(tested_str, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(T, chr);
+
+	sprintf(tested_str2, "%s", tstr);
+	create_file(std_path, tested_str2);
+	rd = fopen(std_path, "r");
+
+	TEST_ASSERT_EQUAL_INT(1, vfscand_wrapper(rd, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(tstr[0], chr);
+
+	rewind(rd);
+
+	TEST_ASSERT_EQUAL_INT(1, fscanf(rd, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(tstr[0], chr);
+
+	TEST_ASSERT_EQUAL_INT(1, vsscanf_wrapper(tested_str, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(tstr[0], chr);
+
+	TEST_ASSERT_EQUAL_INT(1, sscanf(tested_str, "%c", &chr));
+	TEST_ASSERT_EQUAL_CHAR(tstr[0], chr);
+}
+
+TEST(stdio_scanf_str_chr_ptr, str_path)
+{
+
+	char tested_str[50];
+	char chr[50];
+
+	sprintf(tested_str, "%s", std_path);
+	create_file(std_path, tested_str);
+	rd = fopen(std_path, "r");
+
+	TEST_ASSERT_EQUAL_INT(1, vfscand_wrapper(rd, "%s", chr));
+	TEST_ASSERT_EQUAL_STRING(std_path, chr);
+
+	rewind(rd);
+
+	TEST_ASSERT_EQUAL_INT(1, fscanf(rd, "%s", chr));
+	TEST_ASSERT_EQUAL_STRING(std_path, chr);
+
+	TEST_ASSERT_EQUAL_INT(1, vsscanf_wrapper(tested_str, "%s", chr));
+	TEST_ASSERT_EQUAL_STRING(std_path, chr);
+
+	TEST_ASSERT_EQUAL_INT(1, sscanf(tested_str, "%s", chr));
+	TEST_ASSERT_EQUAL_STRING(std_path, chr);
+}
+
+TEST(stdio_scanf_str_chr_ptr, str_special)
+{
+
+	char tested_str[50];
+	char chr[50];
+	char spc[] = "\4399\0ns";
```

Comment:
I doubt if `#` is a special character. Using `\43` is just different way of passing a character value (without an actual character. 

When it comes to the `\0` character it just obscures the next part of the string but we really don`t check it. We don`t know it the `ns` was read or not because by checking it with `TEST_ASSERT_EQUAL_STRING` we only assess the part before `\0` character. We don't check what happened after that, so `ns` is just not used. The idea of something after `\0` is good. Maybe assert equality of arrays, not strings?